tests/security/static/queries/cpp/openenclave/ECall-Args-TOCTOU.ql

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+/**
+ * @name TOCTOU in ECall Aruguments
+ * @description Potential time-of-check and time-of-use vulnerablility in the usage of ECall argiment pointer.
+ * @kind problem
+ * @id ecall-args-toctou
+ * @problem.severity error
+ * @tags security
+ * @precision medium
+ */
+
+import cpp
+import semmle.code.cpp.Type
+import OpenEnclave
+
+from
+  UntrustedMemory hostMem, ECallInputParameter inParam, AssignExpr assExp, VariableAccess va,
+  StackVariable var, Expr expr, FieldAccess fa
+where
+  hostMem.isOriginatedFrom(inParam) and
```

Comment:
Can you describe how this code detects TOCTOU?

Reply:
Its not a generic toctou query. This query is very much specific to detect toctou of host (arg_in) pointer access in ECALL handlers tainting from _handle_ecall to oe_handle_call_enclave_function. 

We have local copy pattern to guard toctou:
> // Copy args to enclave memory to avoid TOCTOU issues.
    args_ptr = (oe_call_enclave_function_args_t*)arg_in;
    args = *args_ptr;

If the host pointer (args_ptr) is accessed directly instead of local copy, this query will catch it.
For example:
        ecall_table.ecalls = _ecall_tables[args_ptr->table_id].ecalls;	
        ecall_table.num_ecalls = _ecall_tables[args_ptr->table_id].num_ecalls;

https://github.com/openenclave/openenclave/commit/0af10738c94f2f9e8d6668767bc60aa0927cc88d#diff-ca075fbde1bf2e0da2cddee75f302504d3f5d893764d262a75be8196f08422cc

Reply:
UntrustedMemory is implemented in OpenEnclave.qll library